tests/EnforcerTest.php

```diff
@@ -428,4 +429,68 @@ public function testMatcherUsingInOperatorBracket()
+            return BuiltinOperations::regexMatch($key1, $key2);
+        });
+
+        $this->assertTrue($e->getRoleManager()->hasLink('u1', 'g1'));
+        $this->assertFalse($e->getRoleManager()->hasLink('u1', 'g2'));
+        $this->assertTrue($e->getRoleManager()->hasLink('u1', 'root'));
+
+        $this->assertTrue($e->enforce('u1', 'data1', 'read'));
+        $this->assertFalse($e->enforce('u2', 'data1', 'read'));
+        $this->assertFalse($e->enforce('u3', 'data2', 'read'));
+        $this->assertFalse($e->enforce('u4', 'data2', 'read'));
+    }
+
+    public function testRbacWithDomain()
+    {
+        $e = new Enforcer(
+            $this->modelAndPolicyPath . '/rbac_with_domain_pattern_model_and_keymatch_model.conf',
+            $this->modelAndPolicyPath . '/rbac_with_domain_pattern_model_and_keymatch_policy.csv'
+        );
+
+        $e->addNamedDomainMatchingFunc('g', 'keyMatch', function (string $key1, string $key2) {
+            return BuiltinOperations::keyMatch($key1, $key2);
+        });
+
+        $this->assertTrue($e->enforce('alice', 'domain1', 'data1', 'read'));
+        $this->assertTrue($e->enforce('alice', 'domain1', 'data1', 'write'));
+        $this->assertTrue($e->enforce('alice', 'domain1', 'data2', 'read'));
+        $this->assertTrue($e->enforce('alice', 'domain1', 'data2', 'write'));
+
+        $this->assertFalse($e->enforce('bob', 'domain1', 'data2', 'write'));
+
+        $this->assertTrue($e->enforce('alice',  'domain2', 'data1', 'read'));
+        $this->assertFalse($e->enforce('alice', 'domain2', 'data1', 'write'));
+        $this->assertTrue($e->enforce('alice',  'domain2', 'data2', 'read'));
+        $this->assertFalse($e->enforce('alice', 'domain2', 'data2', 'write'));
+
+        $this->assertFalse($e->enforce('alice', 'domain3', 'data1', 'read'));
+        $this->assertFalse($e->enforce('alice', 'domain3', 'data1', 'write'));
+        $this->assertFalse($e->enforce('alice', 'domain3', 'data2', 'read'));
+        $this->assertFalse($e->enforce('alice', 'domain3', 'data2', 'write'));
+
+        $this->assertFalse($e->enforce('badr', 'domain4', 'data1', 'write'));
+        $this->assertTrue($e->enforce('badr', 'domain4', 'data1', 'read'));
+        $this->assertTrue($e->enforce('badr', 'domain4', 'data2', 'read'));
+        $this->assertTrue($e->enforce('badr', 'domain4', 'data2', 'write'));
+
+        $this->assertTrue($e->enforce('stef', 'domain1', 'data1', 'read'));
+        $this->assertTrue($e->enforce('stef', 'domain1', 'data1', 'write'));
+        $this->assertTrue($e->enforce('stef', 'domain1', 'data2', 'read'));
+        $this->assertTrue($e->enforce('stef', 'domain1', 'data2', 'write'));
+
+        $this->assertFalse($e->enforce('stef', 'domain5', 'data3', 'read'));
+        $this->assertFalse($e->enforce('ben', 'domain5', 'data3', 'read'));
+        $this->assertTrue($e->enforce('leo', 'domain5', 'data3', 'read'));
+    }
```

Comment:
I re write to be more maintainable & simple by usage of DataProvider.

```suggestion
    /**
     * Data provider for testRbacWithRegexPatternMatching.
     */
    public function regexPatternMatchingDataProvider()
    {
        yield 'user1 with group1' => ['user1', 'group1', true];
        yield 'user1 with group2' => ['user1', 'group2', false];
        yield 'user1 with root' => ['user1', 'root', true];
        yield 'user2 with data1 read' => ['user2', 'data1', 'read', false];
        yield 'user3 with data2 read' => ['user3', 'data2', 'read', false];
        yield 'user4 with data2 read' => ['user4', 'data2', 'read', false];
    }

    /**
     * @dataProvider regexPatternMatchingDataProvider
     */
    public function testRbacWithRegexPatternMatching($user, $role, $expected)
    {
        $enforcer = new Enforcer(
            $this->modelAndPolicyPath . '/rbac_with_regex_pattern_model.conf',
            $this->modelAndPolicyPath . '/rbac_with_regex_pattern_policy.csv'
        );

        $enforcer->addNamedMatchingFunc('g', 'regexMatch', function (string $keyOne, string $keyTwo) {
            return BuiltinOperations::regexMatch($keyOne, $keyTwo);
        });

        $this->assertSame($expected, $enforcer->getRoleManager()->hasLink($user, $role));
    }

    /**
     * Data provider for testRbacWithDomain.
     */
    public function domainDataProvider()
    {
        yield 'alice in domain1 data1 read' => ['alice', 'domain1', 'data1', 'read', true];
        yield 'alice in domain1 data1 write' => ['alice', 'domain1', 'data1', 'write', true];
        yield 'alice in domain1 data2 read' => ['alice', 'domain1', 'data2', 'read', true];
        yield 'alice in domain1 data2 write' => ['alice', 'domain1', 'data2', 'write', true];
        yield 'bob in domain1 data2 write' => ['bob', 'domain1', 'data2', 'write', false];
        yield 'alice in domain2 data1 read' => ['alice', 'domain2', 'data1', 'read', true];
        yield 'alice in domain2 data1 write' => ['alice', 'domain2', 'data1', 'write', false];
        yield 'alice in domain2 data2 read' => ['alice', 'domain2', 'data2', 'read', true];
        yield 'alice in domain2 data2 write' => ['alice', 'domain2', 'data2', 'write', false];
        yield 'alice in domain3 data1 read' => ['alice', 'domain3', 'data1', 'read', false];
        yield 'alice in domain3 data1 write' => ['alice', 'domain3', 'data1', 'write', false];
        yield 'alice in domain3 data2 read' => ['alice', 'domain3', 'data2', 'read', false];
        yield 'alice in domain3 data2 write' => ['alice', 'domain3', 'data2', 'write', false];
        yield 'badr in domain4 data1 write' => ['badr', 'domain4', 'data1', 'write', false];
        yield 'badr in domain4 data1 read' => ['badr', 'domain4', 'data1', 'read', true];
        yield 'badr in domain4 data2 read' => ['badr', 'domain4', 'data2', 'read', true];
        yield 'badr in domain4 data2 write' => ['badr', 'domain4', 'data2', 'write', true];
        yield 'stef in domain1 data1 read' => ['stef', 'domain1', 'data1', 'read', true];
        yield 'stef in domain1 data1 write' => ['stef', 'domain1', 'data1', 'write', true];
        yield 'stef in domain1 data2 read' => ['stef', 'domain1', 'data2', 'read', true];
        yield 'stef in domain1 data2 write' => ['stef', 'domain1', 'data2', 'write', true];
        yield 'stef in domain5 data3 read' => ['stef', 'domain5', 'data3', 'read', false];
        yield 'ben in domain5 data3 read' => ['ben', 'domain5', 'data3', 'read', false];
        yield 'leo in domain5 data3 read' => ['leo', 'domain5', 'data3', 'read', true];
    }

    /**
     * @dataProvider domainDataProvider
     */
    public function testRbacWithDomain($user, $domain, $resource, $action, $expected)
    {
        $enforcer = new Enforcer(
            $this->modelAndPolicyPath . '/rbac_with_domain_pattern_model_and_keymatch_model.conf',
            $this->modelAndPolicyPath . '/rbac_with_domain_pattern_model_and_keymatch_policy.csv'
        );

        $enforcer->addNamedDomainMatchingFunc('g', 'keyMatch', function (string $keyOne, string $keyTwo) {
            return BuiltinOperations::keyMatch($keyOne, $keyTwo);
        });

        $this->assertSame($expected, $enforcer->enforce($user, $domain, $resource, $action));
    }
```
To be consider using an Helper to store DataSet used in DataProvider in static class for more concise code.